launch/launch/launch_context.py

```diff
@@ -171,13 +182,24 @@ def unregister_event_handler(self, event_handler: BaseEventHandler) -> None:
 
     def emit_event_sync(self, event: Event) -> None:
         """Emit an event synchronously."""
-        self.__logger.debug("emitting event synchronously: '{}'".format(event.name))
-        self._event_queue.put_nowait(event)
+        with self._event_queue_lock:
+            self._event_queue.put_nowait(event)
 
     async def emit_event(self, event: Event) -> None:
         """Emit an event."""
-        self.__logger.debug("emitting event: '{}'".format(event.name))
-        await self._event_queue.put(event)
+        with self._event_queue_lock:
+            self.__logger.debug("emitting event: '{}'".format(event.name))
+            await self._event_queue.put(event)
+
+    def is_event_queue_empty(self) -> bool:
+        with self._event_queue_lock:
+            return self._event_queue.empty()
+
+    async def get_next_event(self):
+        with self._event_queue_lock:
+            if self._event_queue.empty():
```

Comment:
what if [shutdown_when_idle](https://github.com/ros2/launch/blob/657745c315ba6a61984b66f168f9c34b3a7e2108/launch/launch/launch_service.py#L302) was false?

This shouldn't be returning immediately in that case, but awaiting somebody to put a new event in the queue.
Maybe, you can add an extra argument to the method `return_when_empty` or something like that.

Reply:
Regardless of that setting, it will cause a busy loop involving `LaunchService._process_one_event()`.

Reply:
Yeah, I was worried about a busy loop here. I think a better alternative is to somehow add an event to the queue if there are no more futures to wait on and the queue is empty.